### DISCUSSION TITLE
Exposing Level-controlled device as MQTT Fan / modeless ZigBee Fan

### DISCUSSION THREAD

**Discussion by lorenz** [Koenkk/zigbee2mqtt#24399] [Devices]
I made a custom Zigbee fan device. As it has fairly fine control (which higher-level control systems do use) and no reasonable auto mode I implemented it as a level-controlled device (having a onOff and a levelCtrl cluster). But trying to write a converter for an MQTT Fan seems impossible without touching a significant chunk of code outside of the converter itself as the HA discovery and device model in Z2M do not support numeric speeds. I hacked together a converter based on a light, which works well but is obviously a hack.

I looked at the HA MQTT Fan spec and it looks like setting something like speed_range_min: 1 speed_range_max: 255 and no presets should result in a value from 1 to 255 arriving via percentage_command_topic, which can then be forwarded directly to the levelCtrl cluster on the Zigbee side. This would also make it convenient for other automation on the MQTT bus to use.

I need to touch at least https://github.com/Koenkk/zigbee-herdsman-converters/blob/a554f6be1e461e19d03c7bf61f748cfcc10ff86b/src/lib/exposes.ts#L557 and https://github.com/Koenkk/zigbee2mqtt/blame/e06848db8c5ad247c27e959eb85465a30134769f/lib/extension/homeassistant.ts#L932.
Basically before I spend a lot of time on this I'd appreciate some input on how this should be done and how we can support both level-controlled fans as well as classical mode-based fans.

Tagging @Koenkk as you own the project and @cmccambridge as you last made significant modifications to the infrastructure I need to touch. Thanks in advance!

**Comment by cmccambridge** [comment]
That matches my recollection, yes - that Z2M's fan model supports only a small finite set of speeds, not an arbitrary numeric range. Also worth noting that (as I recall, from a couple years ago...) Home Assistant's fan model treats speed-based control as mutually exclusive to mode-based control, and intends that any fan "control" that can be used in parallel with a speed control be modeled as a different HA entity, rather than as a fan "mode". In other words, when HA sets a speed, it expects the mode to be cleared, and vice versa.

I think you'd need to do the necessary Z2M fan model changes to support a numeric speed OR a finite set of speeds, and then from there differentiate the entire generated HA MQTT discovery payload based on model of fan, in order to support the 255-level control you've got on your custom device.

A more disruptive/breaking-change level alternative could changing the fan model and MQTT layer to an arbitrary numeric for *all* Z2M fans, and then implement something like the `speed_range_min..speed_range_max` scheme in "reverse" on the ZHC side, where the arbitrary HA/MQTT numeric value gets mapped back to the closest of the limited/finite set of fan speeds supported by the device... Not sure this is much _less_ code, just puts the code in a different place: Complicating the discovery payload vs complicating the handling of MQTT messages.  This does, however, sound like it would be a breaking change to the interface model... Given that value templates are part of HA discovery protocol, home assistant itself would probably work just fine, but other smart home coordinators might be more deeply aware of the existing Z2M fan model and be broken?

FWIW I _think_ this question may be germane to the Inovelli zigbee fans as well, e.g. VZM36? The ZHC converter appears to be pre-selecting some low/medium/high buckets for the fan speed today, and might also benefit from directly forwarding the raw 255-level control to the smart home coordinator: https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/inovelli.ts#L1782-L1796

**Reply by lorenz** [reply]
Considering the differences would it perhaps be a good idea to have two different devices in Z2M? Rename zhc.Fan to zhc.FanWithMode and then add a zhc.FanWithSpeed (names are not great, I know). That would require less logic to differentiate the two. I don't think I'd want to change over the natively mode-based fans to a speed model as some modes do not directly correspond to a speed and there is no good reason to break their API.

Great find on the Inovelli fans, they look like they have the same clusters as mine, so they would indeed also benefit from this.

**Comment by lorenz** [comment]
https://github.com/Koenkk/zigbee-herdsman-converters/pull/8188
https://github.com/Koenkk/zigbee2mqtt/pull/24483

Initial implementation, I havent' touched the Inovelli fans as I don't have any to test.
